pubtools/tracing.py

```diff
@@ -0,0 +1,123 @@
+"""Instrument Pub task functions.
+
+Usage:
+    @instrument_func()
+      def func():
+          pass
+
+"""
+import functools
+import logging
+import os
+import threading
+
+from opentelemetry import baggage, context, trace
+from opentelemetry.baggage.propagation import W3CBaggagePropagator
+from opentelemetry.exporter.otlp.proto.http.trace_exporter import \
+    OTLPSpanExporter
+from opentelemetry.propagate import set_global_textmap
+from opentelemetry.sdk.resources import SERVICE_NAME, Resource
+from opentelemetry.sdk.trace import TracerProvider
+from opentelemetry.sdk.trace.export import BatchSpanProcessor
+from opentelemetry.trace import Status, StatusCode
+from opentelemetry.trace.propagation.tracecontext import \
+    TraceContextTextMapPropagator
+
+propagator = TraceContextTextMapPropagator()
+baggage_propagator = W3CBaggagePropagator()
+log = logging.getLogger(__name__)
+
+
+class TracingWrapper:
+    """Wrapper class that will wrap all methods of calls with the instrument_tracing decorator."""
+
+    __instance = None
+
+    def __new__(cls):
+        """Create a new instance if one does not exist."""
+        if not os.getenv("PUB_OTEL_TRACING", "").lower() == "true":
+            return None
+
+        if TracingWrapper.__instance is None:
+            log.info("Creating TracingWrapper instance")
+            cls.__instance = super().__new__(cls)
+            otlp_exporter = OTLPSpanExporter(
+                endpoint=f"{os.getenv('OTEL_EXPORTER_OTLP_ENDPOINT')}/v1/traces",
+            )
+            cls.provider = TracerProvider(
+                resource=Resource.create({SERVICE_NAME: os.getenv("OTEL_SERVICE_NAME")})
+            )
+            cls.processor = BatchSpanProcessor(otlp_exporter)
+            cls.provider.add_span_processor(cls.processor)
+            trace.set_tracer_provider(cls.provider)
+            set_global_textmap(propagator)
+            cls.tracer = trace.get_tracer(__name__)
+        return cls.__instance
+
+
+def instrument_func(span_name=None, args_to_attr=False):
```

Comment:
Can you please ensure that this appears in online docs at https://release-engineering.github.io/pubtools/ ?

I think you'll have to write a whole new page for this, similar to the `hooks` page which exists now.  The page can contain the docs for this function and also general info about how to enable this (e.g. how to set the relevant env vars).

Reply:
Added docs for it.